### ISSUE TITLE
#255 Stage A0 execution — A/T training, Tenpai learnability gate, and bounded downstream evaluation

### ISSUE THREAD

**Issue by lisbun** [lisbun/lisjong-arena#262]
Refs #255
Refs #258 — COMPLETE / RETAINED AUGMENTATION QUALIFIED
Refs #259 — COMPLETE / A0 PROTOCOL LOCKED — DOWNSTREAM ENABLED
Refs lisbun/lisjong-project#57

## Purpose

#259 で固定された Stage A0 の final executable lock を **変更せず、そのまま一度だけ実行する**。

本Issueの問いは2段階である。

```text
Question 1
  non-riichi structural-Tenpai auxiliary supervision は、
  simple public-conditional baseline を超える hidden-state signal を
  current flat-BC shared representation に形成できるか？

Question 2
  Question 1 が PASS し、
  かつ #259 が downstream ENABLED と事前固定していた場合、
  そのrepresentation変化は bounded development comparison で
  Policy-strength signalにつながるか？
```

本Issueは protocol 設計Issueではない。

```text
#258
  label-path feasibility

#259
  scientific protocol lock

THIS ISSUE
  exact execution + result classification
```

---

# Entry gate — SATISFIED

#259 completed with:

```text
A0 PROTOCOL LOCKED — DOWNSTREAM ENABLED
```

Final Lock B identity:

```text
a7f1c471a8c2f983c5efab9147651eb48b8788eecb1aa296222e21fd9418e93e
```

Scientific materialization is fixed at retained TRAIN 245..264 + VALIDATION 265..270; protected TEST 271..276 remains unread.

Before any model training, strict-read this exact final #259 Lock B artifact.

Require exact match for all bound identities, including at least:

```text
protocol version / identity
Arena revision
lisjong revision
RiichiEnv / rules revision
source / teacher identity
dataset route
TRAIN / VALIDATION identities + digests
feature / vocabulary / legal-mask identity
Tenpai target / eligibility identity
Baseline 0 / Baseline 1 identity
primary metric semantics
block / interval / PASS rule
3 training seeds
interactive-anchor seed
loss normalization
class treatment
lambda_tenpai
optimizer / batch / epoch / patience contract
checkpoint rule
downstream status
opponent population
downstream seeds / rotations / game budget if ENABLED
protected TEST status
```

Any mismatch:

```text
LOCK MISMATCH
  -> STOP / INVALID
```

Do not repair the lock in this Issue.

---

# Non-goals

Do not:

```text
change lambda_tenpai
change class weighting
change training seeds
replace a failed seed
change checkpoint rule
change Baseline 1
inspect a richer baseline and substitute it
change metric / CI / PASS threshold
change corpus / split
extend TRAIN / VALIDATION data
change opponent population
add evaluation seeds
increase game count after seeing uncertainty
change architecture / optimizer
run Wait / Ron auxiliary training
implement Ron labels
add Oracle Guiding / VLOG / privileged critic
read protected formal TEST
promote a Champion
```

No result-driven rescue.

---

# Phase E0 — execution preflight

Before training:

1. strict-read #259 final lock;
2. verify current revisions / dataset digests / sidecar digests;
3. verify TRAIN / VALIDATION non-overlap;
4. verify #258 technical smoke population is not scientific data;
5. verify protected TEST remains unread;
6. verify privileged hidden truth is absent from serialized 8204 Policy input;
7. verify the three locked training seeds and anchor seed;
8. verify Baseline 0 / Baseline 1 artifacts read back strictly;
9. verify the downstream executable plan fixed by Lock B: mechanism-riichi-defense x3, ordered seeds 37700..37905, 206 paired blocks / 1,648 total games; execute it only if Gate A0.5 PASS.

Produce a preflight artifact.

```text
PREFLIGHT PASS
  -> training may begin

PREFLIGHT FAIL
  -> no training
  -> STOP / INVALID
```

---

# Phase E1 — train A / T under all 3 locked seeds

For every locked training seed `s`:

```text
A_s
  Policy only

T_s
  Policy + locked non-riichi structural-Tenpai auxiliary
```

Within seed `s`:

```text
A_s initial shared encoder bytes
==
T_s initial shared encoder bytes

A_s initial Policy-head bytes
==
T_s initial Policy-head bytes
```

Only the auxiliary branch / auxiliary loss path may differ.

Training must use the exact #259 contract:

```text
8204 public features
128-d shared hidden
802-action Policy head
locked legal-action semantics
locked Policy objective
locked optimizer
locked batch semantics
locked epochs / patience
locked lambda_tenpai
locked class treatment
```

No seed may be removed because its optimization looks poor.

No replacement seed may be added.

---

# Training artifact per arm / seed

For each of the six trained models:

```text
A_seed0
T_seed0
A_seed1
T_seed1
A_seed2
T_seed2
```

record at least:

```text
model identity
training seed
initialization fingerprint
dataset identity
protocol identity
epoch history
TRAIN Policy loss
TRAIN auxiliary loss where applicable
VALIDATION Policy CE by epoch
checkpoint-selected epoch
checkpoint SHA-256
strict-load result
parameter count / architecture identity
```

T arms may record auxiliary validation diagnostics, but those diagnostics must not choose checkpoints.

---

# Checkpoint selection

For every A/T arm independently:

```text
selected checkpoint
  = lowest locked VALIDATION choice-row masked Policy CE
```

Tie behavior must follow #259.

Forbidden:

```text
selecting the epoch with best Tenpai loss
selecting a later epoch because Gate A0.5 improves
selecting a checkpoint after game evaluation
```

After checkpoint selection, freeze that model identity.

---

# Phase E2 — Gate A0.5 execution

Only after all six model checkpoints are frozen, execute the locked Tenpai learnability gate.

Primary eligible cell:

```text
opponent RiichiState.NONE
AND
canonical exact structural-Tenpai target AVAILABLE
```

Canonical target remains:

```text
T[j] = OR_t W[j,t]
```

No target redefinition is permitted.

## Baselines

Evaluate on the same locked VALIDATION eligible cells:

```text
Baseline 0
  TRAIN-only global prevalence
  descriptive only

Baseline 1
  locked TRAIN-only public-conditional baseline
  primary comparator
```

Baseline 1 parameters must be read from the #259 artifact, not refit on VALIDATION.

---

# Primary metric

Use the locked binary log loss / Bernoulli NLL semantics.

For each VALIDATION block `h`:

```text
B_h
  = Baseline-1 mean eligible-cell log loss

T_0,h
T_1,h
T_2,h
  = each T-seed model mean eligible-cell log loss

Delta_h
  = B_h - mean(T_0,h, T_1,h, T_2,h)
```

Zero-eligible-cell blocks follow the exact #259 rule.

Do not alter clipping / missing / numerical behavior.

---

# Gate A0.5 classification

Execute the exact locked rule.

PASS requires both:

```text
1. lower bound of locked paired 95% interval
   for mean(Delta_h) > 0

AND

2. each of T_seed0 / T_seed1 / T_seed2
   has positive overall mean improvement
   vs Baseline 1
```

No secondary metric may override the result.

Outcome:

```text
TENPAI LEARNABILITY PASS
```

or:

```text
TENPAI LEARNABILITY FAIL
```

exactly once.

---

# Required Gate A0.5 diagnostics

Report without changing classification:

```text
eligible cell count
eligible block count
excluded / unavailable reason counts

Baseline 0 log loss
Baseline 1 log loss

per-T-seed log loss
per-T-seed improvement vs Baseline 1

mean Delta
locked interval

global Tenpai prevalence
open vs closed non-riichi subgroup
relative-seat subgroup
public progression / turn-stage subgroup
prediction distribution
calibration diagnostic
riichi-opponent diagnostic if predeclared
```

Optional AUROC / Brier etc. remain descriptive only.

---

# Gate routing

## If TENPAI LEARNABILITY FAIL

Stop Stage A0 execution after producing the completion artifact.

Record:

```text
NO DENSE AUXILIARY LEARNING SIGNAL
```

Do not:

```text
run downstream A-vs-T games
activate Stage A1 automatically
train Wait / Ron
change lambda and retry
change baseline and retry
try a fourth seed
```

Parent #255 returns to explicit `REFORMULATE` decision.

---

## If TENPAI LEARNABILITY PASS

Stage A1 becomes eligible at the parent level.

Then inspect only the already-locked #259 downstream status.

### If:

```text
A0 DOWNSTREAM NOT POWERED
```

do not run interactive A/T games.

Record:

```text
TENPAI REPRESENTATION SIGNAL
POLICY VALUE NOT TESTED DUE TO POWER
```

This is a complete valid Stage A0 result.

### If:

```text
A0 DOWNSTREAM ENABLED
```

only then proceed to Phase E3.

Downstream execution must not begin before Gate A0.5 has been classified as PASS.

---

# Phase E3 — bounded A vs T downstream evaluation

Use only the prelocked interactive-anchor training seed.

Do not select whichever training seed looked best at Gate A0.5.

Focal models:

```text
A_anchor
T_anchor
```

Both use the exact frozen checkpoints selected before Gate A0.5.

Use exactly the #259-bound:

```text
opponent/filler population
policy revisions
ordered development seeds
game mode
seat-rotation semantics
max steps
scientific game count
primary score statistic
interval / classification rule
secondary diagnostics
```

No added games.

No seed extension.

No changed opponent after observing results.

---

# Paired execution semantics

When #259 selected the #252-style paired semantics:

```text
same ordered game seed
  -> same 4 focal-seat rotations for A
  -> same 4 focal-seat rotations for T

per seed block:
  score_A,h = mean focal score across rotations
  score_T,h = mean focal score across rotations

D_h = score_T,h - score_A,h
```

The exact implementation must match the locked protocol.

Do not reuse issue-specific #252 candidate/opponent constants unless #259 explicitly bound them.

---

# Downstream primary result

Compute only the locked primary Policy-strength statistic and classification rule from #259.

Report:

```text
block count
A mean score
T mean score
paired mean delta
locked interval
locked classification
```

Do not invent a new threshold or reinterpret the result in this Issue.

Do not turn a favorable secondary metric into a strength success.

---

# Secondary mechanism diagnostics

If predeclared, report:

```text
deal-in rate
deal-in loss
win rate
other existing raw observables explicitly locked by #259
```

Use mechanism-only wording.

Examples:

```text
DEAL-IN REDUCTION SIGNAL
WIN-RATE REDUCTION SIGNAL
NO MECHANISM SIGNAL
```

These must not redefine the primary strength conclusion.

---

# Runtime / execution failures

Do not silently substitute scientific data after execution begins.

For:

```text
training crash
checkpoint corruption
artifact mismatch
game-generation failure
missing locked opponent
seed execution failure
```

follow only the technical-retry rule, if any, explicitly allowed by #259.

If no such retry is locked:

```text
fail closed
record failure
STOP / INVALID
```

A rerun may never be justified by an unfavorable result.

---

# Result exposure / rerun boundary

After the first scientific VALIDATION Gate A0.5 result is exposed:

```text
no model retraining
no seed replacement
no lambda change
no baseline modification
no extra VALIDATION data
no alternate checkpoint
```

After the first downstream score result is exposed:

```text
no extra ordered seeds
no alternate opponent population
no fidelity increase
no second development run to "confirm"
```

Any new experiment requires a new explicit protocol decision / child Issue.

---

# Completion artifact

Produce one strict versioned result artifact binding:

```text
#255 / #258 / #259 / Child 3 identities

protocol identity
execution revision
dataset / sidecar digests
Baseline identities

all 6 trained model identities
initialization fingerprints
checkpoint identities

Gate A0.5:
  eligible counts
  per-block data identity
  Baseline metrics
  per-seed T metrics
  Delta statistics
  interval
  PASS / FAIL

downstream:
  locked status
  NOT RUN / RUN
  exact anchor model
  opponent population
  ordered seeds
  game artifact identities
  paired score result
  interval / classification
  mechanism diagnostics

protected TEST evaluated = false
```

Strict readback required.

---

# Valid final outcomes

Depending on the locked path and results, exactly one primary completion outcome:

```text
NO DENSE AUXILIARY LEARNING SIGNAL
```

or:

```text
TENPAI REPRESENTATION SIGNAL / POLICY VALUE NOT TESTED DUE TO POWER
```

or:

```text
TENPAI REPRESENTATION SIGNAL / POLICY VALUE NOT ESTABLISHED
```

or:

```text
TENPAI SUPERVISION VALUE SIGNAL
```

or, on scientific invalidation:

```text
STOP / INVALID
```

Secondary mechanism wording may accompany but never replace the primary outcome.

---

# Interpretation boundary

A PASS establishes only that:

```text
under this locked corpus
x flat-BC 8204 -> 128 -> 802 family
x this Tenpai target
x this auxiliary loss
x this optimization protocol
```

the learned representation contains a Tenpai signal stronger than the locked simple public-conditional baseline.

It does not prove:

```text
general HandBelief accuracy
Wait prediction quality
Ron-risk prediction quality
explicit HandBelief runtime value
Oracle Guiding value
VLOG value
privileged critic value
Mortal-like architecture value
formal game-strength superiority
```

unless the separately locked downstream evaluation establishes its own corresponding evidence.

A FAIL is equally narrow.

---

# Acceptance criteria

- [ ] #259 final Lock B strict-read succeeds before any model training.
- [ ] All bound revisions / dataset / sidecar identities match exactly.
- [ ] Protected TEST remains unread.
- [ ] Exactly 3 locked training seeds are used.
- [ ] A/T initialization is byte-identical within every seed.
- [ ] No seed is replaced or selected after results.
- [ ] All six model artifacts are provenance-bound.
- [ ] Checkpoints are selected only by locked Policy VALIDATION CE.
- [ ] Gate A0.5 is executed exactly with locked Baseline 1 / metric / block / interval / PASS rule.
- [ ] PASS / FAIL is reported once with no rescue run.
- [ ] No downstream execution begins before Gate A0.5 PASS is known.
- [ ] FAIL never triggers downstream game evaluation.
- [ ] PASS + `NOT POWERED` never triggers downstream game evaluation.
- [ ] PASS + `ENABLED` uses only the locked anchor seed, opponent population, ordered seeds, rotations, and game budget.
- [ ] Downstream conclusion uses only the #259 locked primary classification rule.
- [ ] Secondary mechanism metrics do not override primary strength classification.
- [ ] No Wait / Ron / Oracle Guiding / VLOG work is added.
- [ ] Result artifact strict-read passes.
- [ ] `protected_test_evaluated=false`.
- [ ] One valid completion outcome is recorded.

---

# Handoff after completion

```text
TENPAI LEARNABILITY FAIL
  -> parent #255 = REFORMULATE
  -> do not activate Stage A1 automatically

TENPAI LEARNABILITY PASS
  -> Stage A1 child becomes eligible
  -> eligibility does not depend on downstream strength result

PASS + downstream positive
  -> still do not promote Champion here

PASS + downstream null/negative
  -> still does not reject Wait / Ron auxiliary hypotheses
```

This Issue does not create or execute Stage A1.

**Comment by lisbun** [comment]
## Scheduling status — 2026-09-19 JST

Scientific entry gate remains satisfied and the exact #259 Lock B is ready to consume.

No #262 execution has started yet.

The operator-local environment is currently occupied by the already locked #270 independent confirmation (17,600 games). To avoid resource interference between scientific runs, #262 should remain queued until #270 finishes or the operator explicitly moves it to an independent execution environment.

This is an operational scheduling constraint only. It does not change #262's scientific protocol, Lock B, seeds, downstream enablement, or result-classification contract.

**Comment by lisbun** [comment]
## 2026-09-20 anti-staleness note — locked comparator vs current Champion

Since this execution Issue was locked, the heuristic line has advanced. Arena #297 completed the exact Champion + HandValue-v2 screen as:

```text
CHAMPION + HAND VALUE V2 INCONCLUSIVE
```

with no automatic or manual Champion replacement. The current Heuristic Champion remains:

```text
TargetedHonorReleaseTerminalProgressionPolicy
```

However, #262 must continue to consume #259 Lock B exactly. Its downstream

```text
mechanism-riichi-defense x3
```

population is a **frozen experimental comparator**, not a mutable "current Champion" alias.

Therefore no protocol field, opponent population, seed range, or result interpretation in #262 is changed by #297.

**Comment by lisbun** [comment]
Implementation merged via PR #306.

Merged revision:

```text
0a34ee9b6ef14e1e898df341df1610570764106d
```

Validation on PR head `1b149f1102d814f0de63b878ff0eabc1082a0aef`:
- quality: PASS
- format / Ruff / internal VCS identity: PASS
- 4,252 non-ML tests: PASS (467 skipped)
- ml-tests: PASS
- PR mergeable before merge; no unresolved review comments

Scientific execution has **not** started. No A/T training, Gate A0.5 exposure, downstream A-vs-T games, or protected TEST read has occurred.

#262 remains open. The next step is the one-shot operator-local execution from merged main, using the dedicated locked environment and the exact #259 Lock B / retained corpus / scientific sidecar / public-key artifacts.

**Comment by lisbun** [comment]
Stage A0 one-shot execution was started from merged execution revision `0a34ee9b6ef14e1e898df341df1610570764106d` under the locked downstream runtime:

- lisjong: `f29d129c67e5232d06563c6e457754377734ed14`
- lisjong-engine: `8735e89e1aea000ab59368d0368d476787827741`
- RiichiEnv: `0.4.10`
- Python: `3.14.6`

Observed execution:

```text
PREFLIGHT PASS
scientific rows: TRAIN=9116 VALIDATION=2555
primary_outcome: STOP / INVALID
failure.stage: E1 A/T training
failure.error_type: StageA0ExecutionError
failure.message: A/T initial Policy state fingerprint drifted
protected_test_evaluated: false
result_identity: 760c7e81a4dfe47d886c55755319f2b5304d8b6bdec1eefb9452126e2c68495e
```

No Gate A0.5 result was produced and no downstream games were run. Protected TEST remained unread.

Initial root-cause inspection points to the execution implementation's fingerprint check: it hashes bytes produced by `torch.save(state_dict, BytesIO())` and compares that byte-level serialization across independently materialized `state_dict` objects after strict load. PyTorch serialization bytes are not an appropriate canonical equality contract for tensor content, so this check can report drift even when parameter tensors are value-identical. The merged ML test suite covered checkpoint save/load serving but did not execute `train_seed_pair()`, so this exact runtime boundary was not exercised.

Per #262's locked runtime-failure rule, there is no authorized retry inside this Issue. This execution therefore completes as `STOP / INVALID`. A separate explicit follow-up should audit/fix the fingerprint contract and decide whether a fresh execution is scientifically admissible, noting that no validation Gate result or downstream result was exposed.
